File: modules/expo-auto-attendance/android/src/main/java/expo/modules/autoattendance/GeofenceStore.kt
package expo.modules.autoattendance

import android.content.Context
import android.content.SharedPreferences
import org.json.JSONObject

/**
 * SharedPreferences-backed persistence for the active geofence configuration and
 * the last received transition event.
 *
 * Persisting the configuration lets [GeofenceBootReceiver] re-register the geofence
 * after a reboot or app update, and lets JS query monitoring state at any time —
 * including events delivered while the React context was not alive.
 */
object GeofenceStore {
  private const val PREFS_NAME = "expo_auto_attendance"
  private const val KEY_MONITORING = "monitoring"
  private const val KEY_GEOFENCE = "geofence"
  private const val KEY_LAST_EVENT = "last_event"

  private fun prefs(context: Context): SharedPreferences =
    context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

  fun saveGeofence(
    context: Context,
    identifier: String,
    latitude: Double,
    longitude: Double,
    radius: Float,
  ) {
    val json = JSONObject()
      .put("identifier", identifier)
      .put("latitude", latitude)
      .put("longitude", longitude)
      .put("radius", radius.toDouble())
    prefs(context).edit()
      .putString(KEY_GEOFENCE, json.toString())
      .putBoolean(KEY_MONITORING, true)
      .apply()
  }

  fun clearGeofence(context: Context) {
    prefs(context).edit()
      .remove(KEY_GEOFENCE)
      .putBoolean(KEY_MONITORING, false)
      .apply()
  }

  fun isMonitoring(context: Context): Boolean =
    prefs(context).getBoolean(KEY_MONITORING, false)

  fun getGeofence(context: Context): JSONObject? =
    prefs(context).getString(KEY_GEOFENCE, null)?.let {
      runCatching { JSONObject(it) }.getOrNull()
    }

  fun saveLastEvent(context: Context, transition: String, identifier: String, timestamp: Long) {
    val json = JSONObject()
      .put("transition", transition)
      .put("identifier", identifier)
      .put("timestamp", timestamp)
    prefs(context).edit().putString(KEY_LAST_EVENT, json.toString()).apply()
  }

  fun getLastEvent(context: Context): JSONObject? =
    prefs(context).getString(KEY_LAST_EVENT, null)?.let {
      runCatching { JSONObject(it) }.getOrNull()
    }

  fun clearLastEvent(context: Context) {
    prefs(context).edit().remove(KEY_LAST_EVENT).apply()
  }
}

File: modules/expo-auto-attendance/ios/GeofenceStore.swift
import Foundation

/**
 UserDefaults-backed persistence for the active geofence configuration and the
 last received transition event — the iOS counterpart of the Android
 `GeofenceStore` (SharedPreferences).

 Persisting the last event lets JS read transitions that were delivered while the
 app was relaunched in the background by Core Location (no React context alive).
 */
final class GeofenceStore {
  static let shared = GeofenceStore()

  private let monitoringKey = "expoAutoAttendance.monitoring"
  private let geofenceKey = "expoAutoAttendance.geofence"
  private let lastEventKey = "expoAutoAttendance.lastEvent"

  private let defaults = UserDefaults.standard

  private init() {}

  var isMonitoring: Bool {
    return defaults.bool(forKey: monitoringKey)
  }

  func saveGeofence(identifier: String, latitude: Double, longitude: Double, radius: Double) {
    let geofence: [String: Any] = [
      "identifier": identifier,
      "latitude": latitude,
      "longitude": longitude,
      "radius": radius
    ]
    defaults.set(geofence, forKey: geofenceKey)
    defaults.set(true, forKey: monitoringKey)
  }

  func clearGeofence() {
    defaults.removeObject(forKey: geofenceKey)
    defaults.set(false, forKey: monitoringKey)
  }

  func getGeofence() -> [String: Any]? {
    return defaults.dictionary(forKey: geofenceKey)
  }

  func saveLastEvent(transition: String, identifier: String, timestamp: Int64) {
    let event: [String: Any] = [
      "transition": transition,
      "identifier": identifier,
      "timestamp": timestamp
    ]
    defaults.set(event, forKey: lastEventKey)
  }

  func getLastEvent() -> [String: Any]? {
    return defaults.dictionary(forKey: lastEventKey)
  }

  func clearLastEvent() {
    defaults.removeObject(forKey: lastEventKey)
  }
}
